### ISSUE TITLE
Build config assumes app sources are in current working directory

### ISSUE THREAD

**Issue by halvards** [google/ko#422]
When specifying a [build config](https://github.com/google/ko/blob/780c2812926cd706423e2ba65aeb1beb842c04af/README.md#overriding-go-build-settings) using `.ko.yaml`, the app sources must be in the current working directory.

Adding a [`dir` field](https://github.com/google/ko/blob/780c2812926cd706423e2ba65aeb1beb842c04af/pkg/build/config.go#L71) to the build config to specify the directory containing either `go.mod` or Go source files doesn't result in a successful build.

Initial investigation suggests that the problem may come from using `"."` as the input to [`createBuildConfigs()`](https://github.com/google/ko/blob/780c2812926cd706423e2ba65aeb1beb842c04af/pkg/commands/config.go#L253).

This parameter becomes `baseDir` variable that's used to load module information using [`packages.Load()`](https://github.com/google/ko/blob/780c2812926cd706423e2ba65aeb1beb842c04af/pkg/commands/config.go#L199).

Reproduction steps using ko built from [commit 780c2812926cd706423e2ba65aeb1beb842c04af](https://github.com/google/ko/tree/780c2812926cd706423e2ba65aeb1beb842c04af):

Setup:

```sh
kodir=$(mktemp -d)
git clone https://github.com/google/ko.git $kodir
cd $kodir
git checkout 780c2812926cd706423e2ba65aeb1beb842c04af
go build -o ko .

appdir=$(mktemp -d)
cd $appdir
mkdir -p ./app/cmd/foo
cat << EOF > ./app/cmd/foo/main.go
package main

func main() {
        println("cmd/foo")
}
EOF
cat << EOF > ./app/go.mod
module example.com/foo

go 1.15
EOF
cat << EOF > .ko.yaml
builds:
- id: foo
  dir: ./app
  main: ./cmd/foo
EOF
```

Run:

```sh
$kodir/ko publish --local .
```

Output:

```terminal
2021/08/19 18:57:53 'builds': entry #0 does not contain a usuable path (./app/cmd/foo): /usr/bin/gopackagesdriver: exit status 1: gopkgsdriver: 11.02508ms for GOROOT=/usr/lib/google-golang GOPATH=/usr/local/google/home/halvard/go GO111MODULE=off GOPROXY= PWD=/tmp/tmp.gYb3L3l1wT go list -e -f {{context.ReleaseTags}} -- unsafe
gopkgsdriver: 9.956159ms for GOROOT=/usr/lib/google-golang GOPATH=/usr/local/google/home/halvard/go GO111MODULE= GOPROXY= PWD=/tmp/tmp.gYb3L3l1wT go list -e -json -compiled=false -test=false -export=false -deps=false -find=true -- ./app/cmd/foo
gopkgsdriver: err: exit status 1: stderr: go: go.mod file not found in current directory or any parent directory; see 'go help modules'
```

**Comment by halvards** [comment]
Layout:

```sh
tree -a
```

```terminal
.
├── app
│   ├── cmd
│   │   └── foo
│   │       └── main.go
│   └── go.mod
└── .ko.yaml

3 directories, 3 files
```

**Comment by halvards** [comment]
I'm happy to work on this, so please assign me to this issue.

**Comment by halvards** [comment]
Reopening this issue, as the problem runs deeper.

1. When we look up module information, [we use the working directory](https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/gobuild.go#L221).

2. When we create a `buildContext`, [we set the directory to the working directory](https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/gobuild.go#L238).

3. When we qualify local import paths, [we use the working directory](https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/gobuild.go#L264). The working directory was [set on the `gobuild` instance](https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/gobuild.go#L132), [via the `gobuildOpener` instance](https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/gobuild.go#L250).

In all cases, we don't take into account [the `Dir` field from the build config](https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/config.go#L72).

There can be multiple build configs, and one call to `ko publish` can build multiple images from different targets, so we can't just join working directory + build config directory.

I haven't yet found a way to solve this that avoids breaking changes to exported functions, so I'm opening this to get a discussion started. 

As an aside, this isn't an issue for the Skaffold ko integration, since the Skaffold ko builder only builds one target. This means that the ko `WorkingDirectory` and build config `Dir` can be set to the same value (working dir + build config dir).

**Comment by halvards** [comment]
Some options - not an exhaustive list, and I'd love to get some input here.

1. Use a separate builder instance per target. I'm not sure what impact that'll have on build speeds. It'll also require changes to the build caching and concurrency limiting behavior.

2. Drop support for a per-build config `Dir` field and only allow one value across all configs. BuildConfig with the `Dir` field is on the `main` branch, but it hasn't been present in any released ko version.

**Comment by jonjohnsonjr** [comment]
Yeah this is tricky. I was hoping we could reconcile these two somehow (https://github.com/google/ko/pull/340#discussion_r634630107) but I'm still not sure of the expected semantics.

>  BuildConfig with the Dir field is on the main branch, but it hasn't been present in any released ko version.

This is also true of `WorkingDirectory`, so we have the flexibility to drop either.

> Some options - not an exhaustive list

What about something like...

`WorkingDir` is the default directory, and `Dir` on a `build.Config` can override that value if set?

We could change the way `gobuild` works such that most of [these fields](https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/gobuild.go#L83-L95) are just behind a map?

Or even something like:

```go
type gobuilds struct {
  // per import path config (module info, build context, working dir)
  builders map[string]gobuild

  // default if there's not a specific build config?
  fallback gobuild

  // any "global" config like labels, ko-specific environment variables
}
```

We'd have `gobuilds` implement `build.Interface` and just dispatch to each `gobuild` as needed, which should make any changes needed pretty minimal (for caching and concurrency limiting compositional stuff).

I think the only thing that would be tricky to implement is local import paths: https://github.com/google/ko/blob/a965b407b218b6620caf10aa15b7457a3a82e8a0/pkg/build/gobuild.go#L278-L280

But maybe that's fine? We could just use the default/current directory for that? I think that's generally only used from the command line... but maybe people use that from yaml? We'd need to know the directory of the yaml file for that to make any sense, which I don't think we do today, so maybe this isn't a problem.
